shared/src/simulation/activities/count-patients.ts

```diff
@@ -0,0 +1,69 @@
+import { IsUUID } from 'class-validator';
+import { groupBy } from 'lodash-es';
+import { IsValue } from '../../utils/validators/is-value';
+import { StrictObject, UUID, uuidValidationOptions } from '../../utils';
+import { getCreate } from '../../models/utils/get-create';
+import { isInSpecificSimulatedRegion } from '../../models/utils/position/position-helpers';
+import { Patient } from '../../models/patient';
+import { sendSimulationEvent } from '../events/utils';
+import { PatientsCountedEvent } from '../events/patients-counted';
+import type { PatientStatus } from '../../models/utils/patient-status';
+import type {
+    SimulationActivity,
+    SimulationActivityState,
+} from './simulation-activity';
+
+export class CountPatientsActivityState implements SimulationActivityState {
+    @IsValue('countPatientsActivity' as const)
+    public readonly type = 'countPatientsActivity';
+
+    @IsUUID(4, uuidValidationOptions)
+    public readonly id: UUID;
+
+    /**
+     * @deprecated Use {@link create} instead
+     */
+    constructor(id: UUID) {
+        this.id = id;
+    }
+
+    static readonly create = getCreate(this);
+}
+
+export const countPatientsActivity: SimulationActivity<CountPatientsActivityState> =
+    {
+        activityState: CountPatientsActivityState,
+        tick(
+            draftState,
+            simulatedRegion,
+            _activityState,
+            _tickInterval,
+            terminate
+        ) {
+            const patients = Object.values(draftState.patients).filter(
+                (patient) =>
+                    isInSpecificSimulatedRegion(patient, simulatedRegion.id)
+            );
+            const patientCount = StrictObject.fromEntries(
+                StrictObject.entries(
+                    groupBy(patients, (patient) =>
+                        Patient.getVisibleStatus(
+                            patient,
+                            draftState.configuration.pretriageEnabled,
+                            draftState.configuration.bluePatientsEnabled
+                        )
+                    )
+                ).map(([visibleStatus, patientsOfStatus]) => [
+                    visibleStatus as PatientStatus,
+                    patientsOfStatus.length,
+                ])
+            );
```

Comment:
`groupBy` does not know which keys/groups are possible. Therefore, for empty groups, the key does not exists in the group dictionary instead of being an empty array.

So `patientCount` is of type `Partial<PatientCount>` or `Partial<ResourceDescription<PatientStatus>>` (`PatientCount` and `ResourceDescription<PatientStatus>` are effectively the same type and I would suggest to reuse the `ResourceDescription`, see my comment at the patients counted event). Assigning this partial value to the non-partial value in the patients counted event works (since the typings of `groupBy` claim that it is not partial), but when accessing this value later on, you could get unexpected `undefined` values...